README.md
# IDS Parser

### WARNING!
This is **beta version** (just for demonstration), not for real work! Author is **not** responsible if something goes wrong ;)

### Official website
http://parser2k19.xyz

### Functions
 - Parses log files of well-known Intrusion Detection Systems (IDS), such as Snort, Dallas Lock, Zeek (ex. Bro), etc.
- Shows diagram of legal, illegal & unknown traffic
- Shows traffic description in human-readable format
- Can export all parsed information to PDF

### Download materials

[Files (.zip)](http://parser2k19.xyz/files/ids-parser-beta.zip)

[Files + node_modules (.zip)](http://parser2k19.xyz/files/ids-parser-beta_node.zip)

[Auto install script (for Ubuntu 16+/Debian 8+)](http://parser2k19.xyz/files/ids-parser-install.sh)

[Auto run script (for Ubuntu 16+/Debian 8+)](http://parser2k19.xyz/files/ids-parser-autorun.sh)
           
[Screenshot](http://parser2k19.xyz/files/scr.png)  

[Demo Log Files - also included in this repository, `demo` folder](http://parser2k19.xyz/files/demo-logs.zip)

### Minimal browser requirements

| Name        | Version           | Platform  |
| ------------- |-------------| ---------------|
| Internet Explorer | 10        | Windows 7 SP1 |
| Edge              | 17        | Windows 10, build 1803 |
| Firefox           | 52.9      | Windows XP SP2/Ubuntu 14.04/Mac OS X 10.9 |
| Chrome            | 49        | Windows XP SP3/Ubuntu 14.04/Mac OS X 10.10 |
| Safari            | 5.1.7     | Windows XP SP3/Mac OS X 10.6 |
| Opera             | 36        | Windows XP SP3/ Ubuntu 16.04 (x64)/Mac OS X 10.10 |
| iOS Safari        | 11        | iOS 11.4 |
| Android Browser   | 67        | Android 6 |
| Chrome for Android | 71       | Android 4.1 |
| Firefox for Android | 64      | Android 4.1 |

### How to start
#### .zip archives (local server)
1. Make sure you have successfully connected to MongoDB database (usually *localhost:27017*).
2. Unzip files where you need. Open terminal there.
3. If you have downloaded archive without *node_modules*, run `npm i` command to download libraries for this app.
4. Run nodemon command. When message *Parser was started, using port 3000* is shown, go to `http://localhost:3000/install` and follow all the instructions, after that go to `http://localhost:3000`.

#### Shell scripts (if you have remote server, VPS, etc) - *Ubuntu 16+/Debian 8+ only*
1. Make sure you have successfully connected to MongoDB database (port is usually *27017*).
2. Put `ids-parser-install.sh` file to your server, run this file as sudo. This file will download and unpack libraries that are needed to run the app. Before running, you should delete packages such as `apache`, `mysql-server`, `php-*`.
3. Put `ids-parser-autorun.sh`, make it run when the OS of your server boots up (use file `/etc/rc.local`)
4. Download and unzip the files of parser, put them to your server root (usually `/var/www/html`). If you have downloaded archive without *node_modules*, run `npm i` command at the server root folder to download libraries for this app.
5. Run nodemon command at the server root folder. When message *Parser was started, using port 3000* is shown, go to `http://remote.ip:3000/install` and follow all the instructions, after that go to `http://remote.ip:3000`.
6. If you had binded a domain to your VPS, consider using [nginx proxy for node.js](https://radiostud.io/configuring-nginx-reverse-proxy-node-js-application/) (to get rid of `:3000` port in the address bar).

### License
[GNU GPL v1.0](https://www.gnu.org/licenses/old-licenses/gpl-1.0.en.html)

### Copyright
&copy; [SIV](https://www.facebook.com/thesiv95), 2019

### Libraries & frameworks used

- [Stolzl-Regular](https://fonts.adobe.com/fonts/stolzl) Copyright &copy; 2015 by Mariya Pigoulevskaya. All rights reserved.

- [Font Awesome](https://fontawesome.com) (free license) &copy; 2017 Fonticons, Inc. - Icons — CC BY 4.0 License, Fonts — SIL OFL 1.1 License, Code — MIT License
- [jQuery 1.9.1](https://code.jquery.com/jquery/) &copy; 2013-2019 The jQuery Foundation - MIT License
- [jQuery File Upload v9.25.1](https://blueimp.github.io/jQuery-File-Upload/) &copy; 2019 Sebastian Tschan - MIT License
- [StyleFix 1.0.3 & PrefixFree 1.0.7](https://leaverou.github.io/prefixfree/) &copy; 2017 Lea Verou - MIT License
- [Modernizr 3.6.0](https://modernizr.com) &copy; 2019 Faruk Ateş, Paul Irish, Alex Sexton, Ryan Seddon, Patrick Kettner, Stu Cox, Richard Herrera - MIT License
- [Moment.js 2.24.0](https://momentjs.com) &copy; 2019 JS Foundation and other contributors - MIT License
- [d3.js 5.9.1](https://d3js.org) &copy; 2019 Mike Bostock - BSD license
- [d3pie 0.2.1](http://d3pie.org) &copy; 2014-2015 Benjamin Keen - MIT License
- [mongodb 2.2.33](https://mongodb.github.io/node-mongodb-native/2.2/) &copy; 2009-2012 Christian Amor Kvalheim, 2012-present MongoDB Contributors - Apache 2.0 License
- [xml2js 0.4.19](https://www.npmjs.com/package/xml2js) &copy; 2017 @Leonidas-from-XIV - MIT License
- [Twig 2.x](https://twig.symfony.com) &copy; 2010-2019 Symfony - BSD License
- [Twig 2.x - version for node.js](https://www.npmjs.com/package/twig)
- [Helmet 3.18.0](https://helmetjs.github.io) &copy; 2019 Adam Baldwin - CC 3.0 License
- [x-xss-protection 1.1.0](https://www.npmjs.com/package/x-xss-protection) &copy; 2019 Evan Hahn - MIT License
- [dont-sniff-mimetype 1.0.0](https://www.npmjs.com/package/dont-sniff-mimetype) &copy; 2019 Evan Hahn - MIT License
- [PDFKit](http://pdfkit.org/) &copy; @devongonett 2019 - MIT license
- [svg-to-pdfkit](https://github.com/alafr/SVG-to-PDFKit) &copy; @alafr 2019 - MIT license

CHANGELOG.md
## v 0.1.5
Added parsing warning message D:

## v 0.1.4
- Fixed page clear button
- Deleted unused file

## v 0.1.3
Minified CSS files for distros

## v 0.1.2
 - Added "Clear the page" button to the "Parsing" page
 - Added this changelog file :) just now :)

## v 0.1.1
Improved PDF module

## v 0.1.0
First beta release

index.js
/***** Инициализация и настройка Приложения ******/

// Сторонние библиотеки
var express = require('express');
var app = express();
const helmet = require('helmet');
const xssFilter = require('x-xss-protection');
const nosniff = require('dont-sniff-mimetype');
const mongodb = require('mongodb');
const MongoClient = mongodb.MongoClient;
const assert = require('assert');
const fs = require('fs');
const twig = require('./node_modules/twig');
var upload = require('jquery-file-upload-middleware');
var bodyParser = require('body-parser');
app.use(bodyParser.json());

// Самописные библиотеки и модули
const Draw = require('./core/draw');
const Pdfgen = require('./core/pdfgen');
const Install = require('./install');

// порт для сервера express 
var port = 3000;

// Настройки шаблонизатора Twig
app.set('views', __dirname + '/views'); // папка с HTML-разметкой
app.set('view engine', 'twig');
app.set('twig options', { 
    strict_variables: false
});

// Папка для статических файлов (то есть стили css, скрипты js, и т.п.)
app.use(express.static('public'));

// Безопасные заголовки
app.use(helmet());
app.use(xssFilter());
app.use(nosniff());

// переменная для модуля пдф
var draw = Draw; 

// Подключение к БД, для загрузки настроек
var loadedSetup = {};

MongoClient.connect('mongodb://localhost:27017/config', function(err, db) {
  if (err) {
    throw err;
  }
  db.collection('setup').find().toArray(function(err, result) {
    if (err) {
      throw err;
    }
	
    loadedSetup.lang = result[0]['lang'];
    loadedSetup.styles = result[0]['styles'];
    // console.log(loadedSetup);
    // Загрузка языковых файлов
    var langFile = fs.readFileSync('lang/lang_' + loadedSetup.lang + '.json');
    var loadedLanguage = JSON.parse(langFile);

    /***** Страницы ******/

    app.get('/eula', function(req, res){
        res.render('eula', {
            html_lang: loadedSetup.lang,
            html_dir: loadedLanguage.html_dir,
            msg_noscript: loadedLanguage.msg_noscript,
            msg_old_browser: loadedLanguage.msg_old_browser,
            msg_too_small: loadedLanguage.msg_too_small,
            title: loadedLanguage.header_eula,
            btn_to_main_page: loadedLanguage.btn_to_main_page,
            styles: loadedSetup.styles
        });
            
    });

    app.get('/', function(req, res){ // Главная

        // Решил убрать загрузку лицухи через базу, пусть пока будет отдельной ссылкой
            res.render('index', {
                // Элементы на странице
                html_lang: loadedSetup.lang,
                html_dir: loadedLanguage.html_dir,
                msg_noscript: loadedLanguage.msg_noscript,
                msg_old_browser: loadedLanguage.msg_old_browser,
                msg_too_small: loadedLanguage.msg_too_small,
                title: loadedLanguage.header_main_page,
                header_parsing: loadedLanguage.header_parsing,
                header_settings: loadedLanguage.header_settings,
                header_help: loadedLanguage.header_help,
                footer_eula_info: loadedLanguage.footer_eula_info,
                // Какой стиль выбран?
                styles: loadedSetup.styles
            });
        
        
    });


    app.get('/parsing', function(req, res){   
        res.render('parsing', {
            html_lang: loadedSetup.lang,
            html_dir: loadedLanguage.html_dir,
            msg_noscript: loadedLanguage.msg_noscript,
            msg_old_browser: loadedLanguage.msg_old_browser,
            msg_too_small: loadedLanguage.msg_too_small,
            msg_parsing_warn: loadedLanguage.msg_parsing_warn,
            title: loadedLanguage.header_parsing,
            btn_to_main_page: loadedLanguage.btn_to_main_page,
            btn_start_parsing: loadedLanguage.btn_start_parsing,
            btn_export_parsing: loadedLanguage.btn_export_parsing,
            btn_clear: loadedLanguage.btn_clear,
            parsing_ids_detected: loadedLanguage.parsing_ids_detected,
            parsing_report_download: loadedLanguage.parsing_report_download,
            parsing_timestamp: loadedLanguage.parsing_timestamp,
            parsing_signature: loadedLanguage.parsing_signature,
            parsing_legal: loadedLanguage.parsing_legal,
            parsing_illegal: loadedLanguage.parsing_illegal,
            parsing_unknown: loadedLanguage.parsing_unknown,
            parsing_conn_total: loadedLanguage.parsing_conn_total,
            parsing_date: loadedLanguage.parsing_date,
            parsing_clear: loadedLanguage.parsing_clear,
            draw: Draw, // объект с нужной информацией из модуля Draw.js
            // Какой стиль выбран?
            styles: loadedSetup.styles
        });

    });

    app.get('/settings', function(req, res){
        
        res.render('settings', {
            html_lang: loadedSetup.lang,
            html_dir: loadedLanguage.html_dir,
            msg_noscript: loadedLanguage.msg_noscript,
            msg_old_browser: loadedLanguage.msg_old_browser,
            msg_too_small: loadedLanguage.msg_too_small,
            msg_changed: loadedLanguage.msg_changed,
            title: loadedLanguage.header_settings,
            settings_language: loadedLanguage.settings_language,
            settings_theme: loadedLanguage.settings_theme,
            btn_to_main_page: loadedLanguage.btn_to_main_page,
            btn_apply_settings: loadedLanguage.btn_apply_settings,
            // Какой стиль выбран?
            styles: loadedSetup.styles
        });
    });

    app.get('/help', function(req, res){
        res.render('help', {
            html_lang: loadedSetup.lang,
            html_dir: loadedLanguage.html_dir,
            msg_noscript: loadedLanguage.msg_noscript,
            msg_old_browser: loadedLanguage.msg_old_browser,
            msg_too_small: loadedLanguage.msg_too_small,
            title: loadedLanguage.header_help,
            help_link_parsing: loadedLanguage.help_link_parsing,
            help_link_settings: loadedLanguage.help_link_settings,
            help_link_sysreq: loadedLanguage.help_link_sysreq,
            help_version: loadedLanguage.help_version,
            help_build_date: loadedLanguage.help_build_date,
            help_license: loadedLanguage.help_license,
            btn_to_main_page: loadedLanguage.btn_to_main_page,
            // Какой стиль выбран?
            styles: loadedSetup.styles

        });
    });

    app.get('/author', function(req, res){
        res.render('author', {
            html_lang: loadedSetup.lang,
            html_dir: loadedLanguage.html_dir,
            msg_noscript: loadedLanguage.msg_noscript,
            msg_old_browser: loadedLanguage.msg_old_browser,
            msg_too_small: loadedLanguage.msg_too_small,
            title: loadedLanguage.header_author,
            author_email: loadedLanguage.author_email,
            author_facebook: loadedLanguage.author_facebook,
            author_twitter: loadedLanguage.author_twitter,
            author_github: loadedLanguage.author_github,
            author_vk: loadedLanguage.author_vk,
            author_diploma: loadedLanguage.author_diploma,
            btn_to_main_page: loadedLanguage.btn_to_main_page,
            // Какой стиль выбран?
            styles: loadedSetup.styles
        })
    });

    // Любая другая неизвестная страница должна возвращать ошибку 404
    app.get('*', function(req, res){
        res.render('404', {
            html_lang: loadedSetup.lang,
            html_dir: loadedLanguage.html_dir,
            msg_noscript: loadedLanguage.msg_noscript,
            msg_old_browser: loadedLanguage.msg_old_browser,
            msg_too_small: loadedLanguage.msg_too_small,
            title: loadedLanguage.header_404,
            go_back: loadedLanguage.go_back,
            // Какой стиль выбран?
            styles: loadedSetup.styles
        })
    });

    db.close();

  });
});


// Применение изменений в настройках
app.post("/applysettings", function(req,res){
    
    var setting = {};
    setting.lang = req.body.lang;
    setting.style = req.body.style;
    // console.log('recieved ' + setting.lang + ' ' + setting.style);
    



    MongoClient.connect('mongodb://localhost:27017/config', function(err, db) {
        db.collection("setup").update({a: 'a'}, {lang: setting.lang, styles: setting.style, a: 'a', eula: true}, (err, options) => {
            if (err) {
                console.log('Setup update error: ' + err);
            } else {
                // console.log('Settings applied');
                db.close();
            }

        });
        
    });
 


});



// Вызов модуля PDFRender.js
app.get('/pdf', function(req, res){
    // Подсчет некоторой информации перед ее передачей
    // Воспользуемся объектом draw (инфа, полученная из БД)

    // Если значения undefined (не распознались), то пишем, что названия неизвестны
    if (isNaN(draw.conn_quantity)) draw.conn_quantity = 1;
    if (draw.ids_name === undefined) draw.ids_name = '?';


    // Строка в журнале
    var textString = draw.ip_src + ':' + draw.port_src + '->' + draw.ip_dest + ':' + draw.port_dest + ' [' + draw.protocol + '] at ' + draw.date_reg + ' ' + draw.time_reg + ' - ' + draw.conn_quantity + ' ' + draw.signatures + ' - ' + draw.status;
    var textStringArray = [];
    
    for (var i = 0; i < draw.conn_quantity; i++){
        textStringArray.push(textString);
    }
    var goodTraffic = 0, badTraffic = 0, unknownTraffic = 0;
    for (var i = 0; i < textStringArray.length; i++){
        if (textStringArray[i].indexOf('good') !== -1) {
            goodTraffic++;
        } else if (textStringArray[i].indexOf('bad') !== -1) {
            badTraffic++;
        } else {
            unknownTraffic++;
        }
    }

    // Передадим цвет диаграммы (пока один)
    // Цвет соответствует цветам css-классов (тип трафика)
    var pieColor;

    if (goodTraffic > 0) pieColor = "#51c123";
    if (badTraffic > 0) pieColor = "#da4e0a";
    if (unknownTraffic > 0) pieColor = "#489fca";

    // console.log('pieColor ' + pieColor);

    // Передаваемая информация
    var data = {
        title: 'IDS Parser Report',
        ids_name: draw.ids_name,
        text: textStringArray,
        pieColor: pieColor,
        traffic: {
            good: {
                quantity: goodTraffic,
                percent: parseInt(goodTraffic / draw.conn_quantity * 100)
            },
            bad: {
                quantity: badTraffic,
                percent: parseInt(badTraffic / draw.conn_quantity * 100)
            },
            unknown: {
                quantity: unknownTraffic,
                percent: parseInt(unknownTraffic / draw.conn_quantity * 100)
            }
        },
        total: draw.conn_quantity
    };
    // Главная и единственная функция в модуле
    Pdfgen.renderReport(data, res);
});

// Страница установки
app.get('/install', function(req, res){
    res.render('install');
});

// Вызов модуля запуска установки
app.get('/startinstall', function(){
    Install.init();
})


// Кнопка Начать Обработку
app.get('/extr123', function(req, res){
        
    var extractor = require('./core/extractor'); // именно сюда, иначе скрипт запускается сразу
    extractor.start();
    

});


// Загрузка файла на сервер
upload.configure({
    uploadDir: __dirname + '/public/uploads', // куда сохранять
    uploadUrl: '/uploads' // какая ссылка
});
 
// Если не пост запрос, то возвращаемся на ту же страницу (т.е.ничего не делаем)
app.get('/upload', function( req, res ){
    res.redirect('/parsing');
});
 
app.put('/upload', function( req, res ){
    res.redirect('/parsing');
});
 
app.delete('/upload', function( req, res ){
    res.redirect('/parsing');
});
 
app.use('/upload', function(req, res, next){ // ссылка для загрузки
    upload.fileHandler({
        uploadDir: function () {
            return __dirname + '/public/uploads/'
        },
        uploadUrl: function () {
            return '/uploads'
        }
    })(req, res, next);
});



/***** Запуск сервера ******/
app.listen(port, function () {
    console.log('Parser was started, using port ' + port);
    console.log('Home directory: ' + __dirname);
    console.log('Please go to /install page if it is the first start of the app!');
  });
